OSX migrate to NSCell based renderer (2nd attempt)

csomor opened the pull request:
the old HITheme renderer does not support dark mode for the disclosure button, we should move to use NSCells for rendering in the future, see #24002, replaces PS #24004 with better history of file renderer.mm

csomor committed:
moving renderer.cpp from carbon to cocoa subdir and rename as .mm

this is for keeping the renderer history and preparing for NSCell based renderer methods

csomor committed:
add NSCell based methods

use wxOSX_USE_NSCELL_RENDERER as switch to compare with old renderer (=0)

csomor commented:
Hi @vadz I've tried to setup things correctly for keeping renderer's history, if you think it's ok, then we could merge it, the graphical glitches I've encountered in propgrid (checkbox' focus rects remaining) are already present in the old implementation.

vadz commented on src/osx/cocoa/renderer.mm:
```diff
@@ -162,13 +192,62 @@ wxRendererNative& wxRendererNative::GetDefault()
+wxRendererMac::wxRendererMac()
+{
+#if wxOSX_USE_NSCELL_RENDERER
+    m_nsPushButtonCell = [[NSButtonCell alloc] initTextCell:@""];
```
Do you think it could be worth to allocate all these cells on demand only, i.e. only when a renderer method using it is called for the first time? Or are cells so cheap to create that it's not worth optimizing this?

I'm just a bit wary of adding more and more things unconditionally done on every application startup, each of them doesn't cost much but sooner or later they're going to add up to something noticeable.

csomor replied:
The CPU on my M1 is 6 msec for the allocating all cells, so at the moment I'd rather leave it like that, if that time is ok

vadz replied:
To be honest, 6ms doesn't seem negligible to me: I've just tested and with the debug build of wxGTK (I was too lazy to rebuild in release, but it should be smaller there) the time-to-first-paint is 60ms, i.e. it's 10% of the total startup time.

If you'd like, I can make the changes to create them on demand, but I'd just like to ask you to test them because I'm not sure how to do it.

csomor replied:
> If you'd like, I can make the changes to create them on demand, but I'd just like to ask you to test them because I'm not sure how to do it.

@vadz I'd very much appreciate that, I'll test it, thanks :-)

vadz replied:
I've pushed my changes now (includes a warning fix I was getting here), please let me know if I did anything stupid. TIA!

vadz reviewed (commented):
Thanks a lot, I'm all for merging this, just please let me know if you think it could be worth to create the cells on demand.

vadz committed:
Fix warnings about wrong WX_IS_MACOS_AVAILABLE() use

It can only be used in the "if" statement, not inside a ternary operator
as the latter resulted in

@available does not guard availability here; use if (@available) instead [-Wunsupported-availability-guard]

warning.

vadz committed:
Create NSCells used by wxRenderer on demand

These objects are relatively expansive to create, so don't bother doing
until they're really needed.

csomor committed:
wrapping NSCell on-demand getters

use #if wxOSX_USE_NSCELL_RENDERER here as well

csomor commented:
@vadz Thanks a lot, I've just added a small change, I'm all for merging this ....
